data_model/src/predicate.rs

```diff
@@ -0,0 +1,852 @@
+//! Predicate-related logic. Should contain predicate-related `impl`s.
+
+#[cfg(not(feature = "std"))]
+use alloc::vec;
+
+use super::*;
+use crate::{IdBox, Name, Value};
+
+/// Predicate combinator enum.
+#[derive(Debug, Clone, Serialize, Deserialize, Encode, Decode, IntoSchema)]
+pub enum PredicateBox {
+    /// Logically `&&` the results of applying the two predicates.
+    And(Vec<PredicateBox>),
+    /// Logically `||` the results of applying the two predicates.
+    Or(Vec<PredicateBox>),
+    /// Negate the result of applying the predicate.
+    Not(Box<PredicateBox>),
+    /// The raw predicate that must be applied.
+    Raw(value::Predicate),
+}
+
+impl PredicateBox {
+    /// Construct [`PredicateBox::Raw`] variant.
+    #[inline]
+    pub fn new(pred: impl Into<value::Predicate>) -> Self {
+        Self::Raw(pred.into())
+    }
+
+    /// Construct [`PredicateBox::And`] variant.
+    #[inline]
+    pub fn and(left: impl Into<PredicateBox>, right: impl Into<PredicateBox>) -> Self {
+        Self::And(vec![left.into(), right.into()])
+    }
+
+    /// Construct [`PredicateBox::Or`] variant.
+    #[inline]
+    pub fn or(left: impl Into<PredicateBox>, right: impl Into<PredicateBox>) -> Self {
+        Self::Or(vec![left.into(), right.into()])
+    }
+
+    /// Convert instance into its negation.
+    #[must_use]
+    #[inline]
+    pub fn negate(self) -> Self {
+        match self {
+            Self::And(preds) => Self::Or(preds.into_iter().map(Self::negate).collect()),
+            Self::Or(preds) => Self::And(preds.into_iter().map(Self::negate).collect()),
+            Self::Not(pred) => *pred, // TODO: should we recursively simplify?
+            Self::Raw(pred) => Self::Not(Box::new(PredicateBox::Raw(pred))),
+        }
+    }
+}
+
+impl Default for PredicateBox {
+    fn default() -> Self {
+        PredicateBox::Raw(value::Predicate::Pass)
+    }
+}
+
+impl PredicateTrait<Value> for PredicateBox {
+    #[inline] // This is not a simple function, but it allows you to inline the logic and optimise away the logical operations.
+    fn applies(&self, input: &Value) -> bool {
+        match self {
+            PredicateBox::And(vector) => vector.iter().all(|pred| pred.applies(input)),
+            PredicateBox::Or(vector) => vector.iter().any(|pred| pred.applies(input)),
+            PredicateBox::Not(predicate) => !predicate.applies(input),
+            PredicateBox::Raw(predicate) => predicate.applies(input),
+        }
+    }
+}
+
+#[cfg(test)]
+#[allow(clippy::print_stdout, clippy::use_debug)]
+pub mod test {
```

Comment:
```suggestion
pub mod test {
    #![allow(clippy::print_stdout, clippy::use_debug)]

```

Reply:
Good thinking, it should help if we decide to extract them into their own files.